### PR TITLE
Add test_long_chain_with_restart_from_snapshot

### PR DESCRIPTION
**Pull request by posvyatokum**
The main motivation for the test is to see how garbage collection will behave without caches if we restart from the middle of an epoch.

### CODE REVIEW

**Comment by posvyatokum** [comment]
Adding test that should fail before the fix mentioned above. But the CI is green.
The fail is supposed to come from `log_assert!` here https://github.com/near/nearcore/blob/4eb4d0dae4b59bcfdde02eed4d827cc1aa85b005/chain/client/src/client.rs#L1542
`log_assert!` only panics in debug mode, and our CI pipelines use `--cargo-profile quick-release`
@Ekleog-NEAR do you have ideas? I really want this to fail per-commit. Also, this test can be much shorter, if I can enable `no_cache` feature in it.

**Comment by Ekleog-NEAR** [comment]
@nagisa I recall we discussed quick-release vs. dev for CI. It seems like quick-release is running without debug assertions? If so, I think we should revert and switch back to using the dev profile for CI, even though it might be slower, because it could miss intentionally-placed debug assertions :/

**Comment by Ekleog-NEAR** [comment]
> Also, this test can be much shorter, if I can enable no_cache feature in it.

I guess you mean the compile-time no_cache? It’s something that we unfortunately can’t enable in a single test, not without making it a real snowflake that will become hard to maintain (unless we have a lot of tests that require this).

Would it be hard to remove the compile-time no_cache feature, and replace it with a runtime configuration, that could be turned on/off from the test itself?

**Comment by nagisa** [comment]
> (Or we could add debug-assertions for quick-release, which might make sense considering it’s not for production usage, but then it’d probably be better to rename it for something like dev-release to be clear it’s not intended for prod)

I’m in favour of making it `dev-release`. My memory is that unoptimized builds will start taking too long to run to fit within even the timeouts we have right now, especially on MacOS.

File: integration-tests/src/tests/client/process_blocks.rs [hunk 1]
```diff
@@ -3431,6 +3431,49 @@ fn test_catchup_no_sharding_change() {
...
+
+    let max_height = epoch_length * env1.clients[0].config.gc.gc_num_epochs_to_keep + 3;
+
+    for h in 1..max_height {
+        let block = env1.clients[0].produce_block(h).unwrap().unwrap();
+        env1.process_block(0, block.clone(), Provenance::PRODUCED);
+    }
+
+    let mut env2 = TestEnv::builder(chain_genesis)
+        .stores(vec![env1.clients[0].chain.store().store().clone()])
```
Comment (wacban) [inline integration-tests/src/tests/client/process_blocks.rs:3465]: neat!

**Review by wacban — APPROVED** [review]
LGTM

**Pull request merged**